# Gouania Phylogenomics

This repository contains all scripts and Jupyter notebooks used for the primary and downstream analyses presented in:

Maximilian Wagner, Philipp Resl, Nadine Klar, Jonathan M Huie, Iliana Bista, Shane McCarthy, Michelle Smith, Richard Durbin, Stephan Koblmüller, Hannes Svardal, **_The genomics of convergentadaptation to intertidal gravel beaches in Mediterranean clingfishes_**, Genome Biology and Evolution, 2026;, evag031, https://doi.org/10.1093/gbe/evag031

The code covers data processing, phylogenomic analyses, and comparative genomic workflows used in the study.

## Data availability

All raw input data and processed results required to run the notebooks are publicly available on Figshare:  
https://doi.org/10.6084/m9.figshare.31238869

The newly generated DNA barcode sequences, raw sequencing files, and variant call sets underlying this article are available from the following repositories:

- **GenBank Nucleotide Database** (https://www.ncbi.nlm.nih.gov/)  
  Accession numbers: **OL839338–OL944591**
- **European Nucleotide Archive (ENA)** (https://www.ebi.ac.uk/ena/browser/home)  
  Project accession number: **PRJEB49819**
- **European Variation Archive (EVA)** (https://www.ebi.ac.uk/eva/)  
  Project accession number: **PRJEB60875**
- **Draft Genome assemblies**
  https://doi.org/10.6084/m9.figshare.31238869

Individual accession numbers are listed in the supplementary material of the manuscript.

## Code availability

The bioinformatic pipelines, written in **Snakemake** (Mölder *et al.* 2021), and the Jupyter notebooks used for local and downstream analyses are available in this repository:  
https://github.com/maxwagn/Gouania-Phylogenomics


